src/vec_bcd_ppc.h

```diff
@@ -554,9 +1400,478 @@ vec_bcdsub (vBCD_t a, vBCD_t b)
+      "v" (vra)
+      : "cr6" );
+#else
+  const vi8_t zero = vec_splat_s8(0);
+  vi8_t shd = vec_splat (vrb, VEC_BYTE_L_DWH);
+  vui128_t t;
+  // Multiply digit shift by 4 to get bit shift count
+  shd = vec_add (shd, shd);
+  shd = vec_add (shd, shd);
+  t = (vui128_t) vra;
+  // Compare shift positive or negative
+  if (vec_all_ge(shd, zero))
+    {
+      // Positive, shift left
+      t = vec_slq (t, (vui128_t) shd);
+    }
+  else
+    {
+      // Negative, shift right by absolute value
+      shd = vec_sub (zero, shd);
+      t = vec_srq (t, (vui128_t) shd);
+    }
+  vrt = (vBCD_t) t;
+#endif
+  return (vrt);
+}
+
+/** \brief Combined Decimal Add & write Carry Signed Quadword.
+ *
+ * Two Signed 31 digit BCD values are added, and the carry-out
+ * (the high order 32nd digit) of the sum is generated.
+ * Botht the sum and the carry are returned.
+ *
+ * |processor|Latency|Throughput|
+ * |--------:|:-----:|:---------|
+ * |power8   | 15-24 | 1/cycle  |
+ * |power9   | 6-15  | 2/cycle  |
+ *
+ * @param *cout a pointer to a 128-bit vector to recieve the BCD carry-out.
+ * @param a a 128-bit vector treated as a signed BCD 31 digit value.
+ * @param b a 128-bit vector treated as a signed BCD 31 digit value.
+ * @return a 128-bit vector with the low order 31-digits of the sum (a+b).
+ * Values are -1, 0, and +1.
+ */
+static inline vBCD_t
+vec_cbcdaddcsq (vBCD_t *cout, vBCD_t a, vBCD_t b)
+{
+  vBCD_t t, c;
+#ifdef _ARCH_PWR8
+  vBCD_t sum_ab, sign_a, sign_ab;
+
+  sum_ab = vec_bcdadd (a, b);
+  if (__builtin_expect (__builtin_bcdadd_ov ((vi128_t) a, (vi128_t) b, 0), 0))
+    {
+      c = vec_bcdcpsgn (_BCD_CONST_PLUS_ONE, sum_ab);
+    }
+  else // (a + b) did not overflow, but did it borrow?
+    {
+      c = _BCD_CONST_ZERO;
+      sign_ab = vec_bcdcpsgn (sum_ab, a);
+      if (!vec_all_eq(sign_ab, sum_ab))
+	{
+	  vBCD_t nines = vec_bcdcpsgn (_BCD_CONST_PLUS_NINES, a);
+	  vBCD_t tensc = vec_bcdcpsgn (_BCD_CONST_PLUS_ONE, a);
+	  c = vec_bcdcpsgn (_BCD_CONST_PLUS_ONE, sum_ab);
+	  sum_ab = vec_bcdaddesqm (nines, sum_ab, tensc);
+	}
+    }
+  t = sum_ab;
+#else
+  vBCD_t sign_ab;
+  _Decimal128 d_a, d_b, d_s, d_t;
+  const vui32_t mz = CONST_VINT128_W(0, 0, 0, 0x0000000d);
+  d_a = vec_BCD2DFP (a);
+  d_b = vec_BCD2DFP (b);
+  d_s = d_a + d_b;
+  t = vec_DFP2BCD (d_s);
+  // Shift right 31 digits, leaving the carry.
+  d_t = __builtin_dscriq (d_s, 31);
+  c = vec_DFP2BCD (d_t);
+  // fix up spurious negative zeros
+  if (vec_all_eq((vui32_t ) c, mz))
+    c = _BCD_CONST_ZERO;
+  // (a + b) did not overflow, but did it borrow?
+  sign_ab = vec_bcdcpsgn (t, a);
+  if (!vec_all_eq(sign_ab, t))
+    {
+      vBCD_t nines = vec_bcdcpsgn (_BCD_CONST_PLUS_NINES, a);
+      vBCD_t tensc = vec_bcdcpsgn (_BCD_CONST_PLUS_ONE, a);
+      c = vec_bcdcpsgn (_BCD_CONST_PLUS_ONE, t);
+      t = vec_bcdaddesqm (nines, t, tensc);
+    }
+#endif
+  *cout = c;
+  return (t);
+}
+
+/** \brief Combined Decimal Add Extended & write Carry Signed Quadword.
```

Comment:
s/write/Write/

Reply:
Ok
But you talked to Brett and convinced him that the PowerISA is wrong about write-Carry and needs to change, right?